File: Learn_Android/Old/onKeyLongPress/app/src/main/java/com/example/user/onkeylongpress/MainActivity.java
package com.example.user.onkeylongpress;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.KeyEvent;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    private boolean shortPress=false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
    }

    @Override
    public boolean onKeyLongPress(int keyCode, KeyEvent event) {
        if(keyCode==KeyEvent.KEYCODE_VOLUME_DOWN){
            shortPress=false;
            Toast.makeText(this,"longPress",Toast.LENGTH_LONG).show();
            return true;
        }
        return false;
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if(keyCode==KeyEvent.KEYCODE_VOLUME_DOWN){
            if(event.getAction()==KeyEvent.ACTION_DOWN){
                event.startTracking();
                if(event.getRepeatCount()==0){
                    shortPress=true;
                }
                return true;
            }
        }
        return super.onKeyDown(keyCode,event);
    }

    @Override
    public boolean onKeyUp(int keyCode, KeyEvent event) {
        if(keyCode==KeyEvent.KEYCODE_VOLUME_DOWN){
            if(shortPress){
                Toast.makeText(this,"shortPress",Toast.LENGTH_LONG).show();
            }
            shortPress=false;
            return true;
        }
        return super.onKeyUp(keyCode,event);
    }
}

File: Learn_Android/Di_Yi_Hang_Dai_Ma(3rd edition)/Retrofit/app/src/main/java/com/example/retrofit/MainActivity.kt
package com.example.retrofit

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Button
import android.widget.Toast
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.lang.Exception

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        findViewById<Button>(R.id.getAppDataBtn).setOnClickListener {
            val retrofit = Retrofit.Builder()
                .baseUrl("https://sunkai.xyz:5001")
                .addConverterFactory(GsonConverterFactory.create())
                .build()
            val appService = retrofit.create(AppService::class.java)
            appService.getAppData().enqueue(object:Callback<List<MainPageBanner>>{
                override fun onResponse(
                    call: Call<List<MainPageBanner>>,
                    response: Response<List<MainPageBanner>>
                ) {
                    val list = response.body()
                    println(list)
                }

                override fun onFailure(call: Call<List<MainPageBanner>>, t: Throwable) {
                    t.printStackTrace()
                }
            })
        }
        findViewById<Button>(R.id.getNewsList).setOnClickListener {
            val appService = ServiceCreator.create<AppService>()
            appService.getNewsList(1).enqueue(object :Callback<List<NewsListResponse>>{
                override fun onResponse(
                    call: Call<List<NewsListResponse>>,
                    response: Response<List<NewsListResponse>>
                ) {
                    Toast.makeText(this@MainActivity, "News list returned", Toast.LENGTH_SHORT).show()
                }

                override fun onFailure(call: Call<List<NewsListResponse>>, t: Throwable) {

                }

            })
        }

        findViewById<Button>(R.id.getNewsDetail).setOnClickListener {

        }
    }

    suspend fun getDetail(){
        try {
            val newsDetail = ServiceCreator.create<AppService>().getNewsDetail("test_id").await()
        }catch (e:Exception){
            e.printStackTrace()
        }
    }
}